renderer.php

```diff
@@ -77,9 +79,10 @@ public function render_envbar($match, $fixed = true, $envs = array()) {
         // If passed a list of env's, then for any env in the list which
         // isn't the one we are on, and which isn't production, add some
         // css which highlights broken links which jump between env's.
-        foreach ($envs as $env) {
-            if ($env->matchpattern != $match->matchpattern) {
-                $css .= <<<EOD
+        if ($config->highlightlinks) {
```

Comment:
If this config value is not set,  it will produce an error like undefined variable. So probably it's better to use get_config('local_envbar', highlightlinks) 
There are many places where the same could happen. I'd suggest to change them all or have a local_envbar_config($name) function where you will check if the value is set before return a value. 
To be honest this part of moode is a bit annoying, when sometimes you don't get default values set in DB.

Reply:
I understand your concerns, however I think this is safe at runtime here. The attributes of the $config object I am processing, like $config->highlightlinks, are all based on admin settings and are thus stored in the DB when the plugin is installed (with their default values or with the values which the admin has directly set at installation time).

The only time when I have encountered an undefined variable problem like you describe was when a $config->foo attribute was processed before the plugin was properly installed, for example within a _extend_navigation() hook function. See https://github.com/moodleuulm/moodle-local_boostnavigation/blob/master/lib.php#L41-L44 for an example.

If it fits your demands better, I will be happy to rewrite everything to individual get_config('local_envbar', highlightlinks) calls, but creating a local_envbar_config($name) function would be overkill IMHO.

What do you think?

Reply:
I have done more tests and looks like the recent moodle manages these cases better. All works as expected when I bumped plugin version and ran upgrade script.